### PR TITLE
Key_Shared subscription on non-persistent topic

### PR DESCRIPTION
**Pull request by codelipenghui**
### Motivation

Support Key_Shared subscription on non-persistent topic

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (o)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: ( no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by codelipenghui** [comment]
relate to #4077

**Comment by jiazhai** [comment]
run cpp tests

**Comment by codelipenghui** [comment]
run java8 tests

**Comment by wolfstudy** [comment]
LGTM +1

run java8 tests

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonPersistentKeySharedSubscriptionTest.java [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+                .create();
+
+        int consumer1Slot = HashRangeStickyKeyConsumerSelector.DEFAULT_RANGE_SIZE;
+        int consumer2Slot = consumer1Slot >> 1;
+        int consumer3Slot = consumer2Slot >> 1;
+
+        int consumer1ExpectMessages = 0;
+        int consumer2ExpectMessages = 0;
+        int consumer3ExpectMessages = 0;
+
```
Comment (wolfstudy) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonPersistentKeySharedSubscriptionTest.java]: In the test case, the code of `newConsumer()` and `newProducer()` is repeated multiple times, whether consider encapsulation as a function?

Comment (codelipenghui) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonPersistentKeySharedSubscriptionTest.java]: Nice suggestion, i'm already and separate method `newConsumer` and `newProducer` to create consumer and producer, PTAL.

Comment (wolfstudy) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/NonPersistentKeySharedSubscriptionTest.java]: LGTM

**Pull request merged**